src/stores/search.js

```diff
@@ -0,0 +1,172 @@
+import { defineStore } from 'pinia'
+import { computed, reactive, readonly, toRefs } from '@nuxtjs/composition-api'
+
+import { ALL_MEDIA } from '~/constants/media'
+import { useFilterStore } from '~/stores/filter'
+
+import {
+  filtersToQueryData,
+  queryStringToSearchType,
+} from '~/utils/search-query-transform'
+
+/**
+ * Store information about the search type and search query.
+ * Search query is based on the filter values, but has important differences:
+ * - parameter names are the names that the API expects.
+ * - some parameter names correspond to different filter names for different media types:
+ * `categories` can mean `imageCategories` if the current `searchType` is `IMAGE`, or `audioCategories`
+ * if it is `AUDIO`.
+ * - query parameter values are the values that are used in the API query.
+ */
+export const useSearchStore = defineStore('search', () => {
+  const filterStore = useFilterStore()
+  /**
+   * @type {import('../store/types').SearchState}}
+   */
+  const state = reactive({
+    searchType: ALL_MEDIA,
+    query: {
+      q: '',
+      license_type: '',
+      license: '',
+      categories: '',
+      extension: '',
+      duration: '',
+      aspect_ratio: '',
+      size: '',
+      source: '',
+      searchBy: '',
+      // mature can be either 'true' or ''
+      mature: '',
+    },
+  })
+  const { searchType, query } = toRefs(state)
+
+  /**
+   * @param {import('../store/types').SupportedSearchType} type
+   */
+  const setSearchType = (type) => {
+    state.searchType = type
+    filterStore.setSearchType(type)
+  }
+
+  /**
+   * Returns the search query parameters for API request:
+   * drops all parameters with blank values.
+   *
+   * @returns {import('@nuxtjs/composition-api').ComputedRef<import('../store/types').ApiQueryParams>}
+   */
+  const searchQueryParams = computed(() => {
+    // Ensure that q filter always comes first
+    return Object.keys(state.query).reduce(
+      (obj, key) => {
+        return !['q'].includes(key) && state.query[key].length
+          ? { ...obj, [key]: state.query[key] }
```

Comment:
I read a blog post recently that changed my perspective on how `reduce` is used. https://www.richsnapp.com/article/2019/06-09-reduce-spread-anti-pattern

I think the post is a _little_ dramatic (I wouldn't call this an "anti-pattern" personally) but the gist is that because the accumulator object is local to the reduce function, it's actually fine to mutate it directly instead of creating a new copy per iteration.

Mostly boils down to just using `Object.assign` if you want a psuedo-functional approach still or plain old property assignment to the accumulator.

```suggestion
          ? Object.assign(obj, { [key]: state.query[key] })
```

I wonder what you think about this? Is it something you've considered before? I personally still like the ergonomics of the spread operator here and find it more readable than `Object.assign`. I'm also skeptical that the performance impact of this is that significant in most use cases (hard to tell for this one though! We have a variable amount of filters to work through). My understanding is that JavaScript engines (or at least V8, I haven't read this about Spidermonkey or CoreJS but I'm assuming they work similarly) create dynamically defined classes under the hood to represent individual JavaScript `Object` instances and that they have to recreate the class definition whenever you add another property.

This makes me wonder what the performance difference between something like `Object.fromEntries(...)` with a `reduce` to generate the array of key/value pairs is vs `reduce`ing to an Object is. Presumably the `fromEntries` version gives the runtime more up-front information about what the resulting object is going to be, but does the engine actually use that information?

Anyway, just some food for thought, not any kind of concrete recommendation :slightly_smiling_face:

Reply:
This is very interesting, and great to know! I have simplified some `reduce`s, and will be careful with them in the future.